securebanking-openbanking-uk-gateway-conversion-filter/src/main/java/com/forgerock/securebanking/uk/gateway/conversion/filter/IntentConverterFilter.java

```diff
@@ -0,0 +1,211 @@
+                printFilterInfo();
+            }
+            String jsonPayload = getEntity(context, request, next);
+
+            Object objectMapped = convert(jsonPayload);
+
+            byte[] serialised = toBytes(objectMapped);
+
+            return processResult(serialised, context, request, next);
+        } catch (Exception e) {
+            logger.error("Conversion to OB Object filter Error\n", e);
+            ResponseException responseException = new ResponseException(e.getMessage(), e);
+            // Overriding the internal server error status to bad request
+            responseException.getResponse().setStatus(Status.BAD_REQUEST);
+            return newResponsePromise(responseException.getResponse());
+        }
+    }
+
+    /** Creates and initializes a ConversionFilter in a heap environment. */
+    public static class Heaplet extends GenericHeaplet {
+
+        @Override
+        public Object create() {
+            IntentType intentType = config.get(CONFIG_FIELD_INTENT_TYPE)
+                    .required()
+                    .as(enumConstant(IntentType.class));
+            logger.trace("intentType = {}", intentType);
+            MessageType payloadFrom = config.get(CONFIG_FIELD_PAYLOAD_FROM)
+                    .required()
+                    .defaultTo(MessageType.REQUEST.toString())
+                    .as(enumConstant(MessageType.class));
+            logger.trace("payload From {}", payloadFrom);
+            List<MessageType> payloadTo = config.get(CONFIG_FIELD_RESULT_TO)
+                    .defaultTo(List.of(MessageType.RESPONSE.toString()))
+                    .as(messageTypeList());
+            return new IntentConverterFilter(intentType, payloadFrom, payloadTo);
+        }
+    }
+
+    private static Function<JsonValue, List<MessageType>, JsonValueException> messageTypeList() {
+        return jsonValue -> {
+            List<String> jsonValueList = jsonValue.asList(String.class);
+            List<MessageType> resultList = new ArrayList<>();
+            for (String value : jsonValueList) {
+                if (!Enums.getIfPresent(MessageType.class, value).isPresent()) {
+                    String message = String.format("Configuration 'resultTo' %s list contains not supported values," +
+                            " all configuration values should be a MessageType values.", jsonValueList );
+                    logger.error(message);
+                    throw new JsonValueException(jsonValue, message);
+                }
+                resultList.add(MessageType.valueOf(value));
+            }
+            return resultList;
+        };
+    }
+
+    private String getEntity(Context context, Request request, Handler next) throws IOException, InterruptedException {
+        logger.trace("Payload from {}", this.payloadFrom);
+        if (payloadFrom.equals(MessageType.RESPONSE)) {
+            return getEntity(next.handle(context, request).getOrThrow());
```

Comment:
Promise.getOrThrow is a blocking call, which is not allowed unfortunately, see: https://backstage.forgerock.com/docs/ig/7.2/gateway-guide/extending.html#about-custom-extensions

Reply:
I don't think this method is needed, you can create a new method which has the logic to go from a Message obj to the converted json byte[].

This new method can then be plugged into the request flow and the response flow (within the promise callback within method processResult)

Reply:
Refactor done.